File: Easy/001-two-sum/javascript/solution.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var twoSum = function(nums, target) {
    let pairs = {}
    for(let i = 0; i < nums.length; i++) {
        const complement = target - nums[i];
        if(pairs[complement] !== undefined) {
            return [pairs[complement], i]
        } else {
            const value = nums[i]
            pairs[value] = i
        }
    }
    return []
};

module.exports = twoSum;

File: Easy/001-two-sum/python/solution.py
"""
Finds two indices such that their values add up to the target.
Time Complexity: O(n)
Space Complexity: O(n)
"""
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        # This dictionary stores: {value: index}
        seen_values = {}
        
        for index, num in enumerate(nums):
            # Calculate the "complement" needed to reach the target
            complement = target - num
            
            # Check if we have already encountered the complement
            if complement in seen_values:
                # If found, return the index of the complement and current index
                return [seen_values[complement], index]
            
            # Otherwise, store the current number and its index for future lookups
            seen_values[num] = index
            
        return [] # Return empty if no solution is found